UnitTest/tests/test_DeviceData_rmsg-definition.txt

```diff
@@ -1,130 +1,135 @@
-	}
     subtest 'Test with pre-release SD_Device_ProtocolList' => sub {
-    	TODO: {
-    		local $TODO = "Checking with prerelease Version of SD_Device_ProtocolList which can fail";
-    	
-		  	loadJson("https://raw.githubusercontent.com/RFD-FHEM/SIGNALduino_TOOL/pre-release/FHEM/lib/SD_Device_ProtocolList.json");
-	
-			my $pID;
-			my $tID;
-			
-			my $SD_Dispatch = $mock->mock("SIGNALduno_Dispatch"); 
-			my $noTestRun=1;
-			use lib::SD_Protocols;
-		    while ( ($pID, $testSet) = each  (@{$testDataArray}) )
-			{
-				next if (!lib::SD_Protocols::exists($testSet->{id}));
-				while ( ($tID, $tData) = each (@{$testSet->{data}}) ) 
-				{
-					next if (!defined($tData->{rmsg}) || !defined($tData->{dmsg}) || !defined($tData->{internals}) );
-					
-			        subtest "[$pID]: $testSet->{name}: [$tID] " => sub {
-			            $SD_Dispatch->reset();
-						$SD_Dispatch->side_effect(\&VerifyDispatch);
-						$id_matched=0;
-						$dmsg_matched=0;
-						$SD_Dispatch_calledCounter=0;
-			            SIGNALduino_Parse($targetHash, $targetHash, $targetHash->{NAME}, "\002".$tData->{rmsg}."\003") if (defined($tData->{rmsg}));
-		           		if ($SD_Dispatch->called() >0 )  {
-		              		$noTestRun=0;
-		
-							ok($id_matched,"SIGNALduno_Dispatch check id ") || note explain (($SD_Dispatch->called_with())[4] , " vs ", $testSet->{id});
-		              		ok($dmsg_matched,"SIGNALduno_Dispatch check dmsg ") || note explain (($SD_Dispatch->called_with())[2] , " vs ", $testSet->{dmsg});
-		
-							my $expected_repeat=min(AttrVal($target,"maxMuMsgRepeat",4),$tData->{dispatch_repeats});
-							is($tData->{dispatch_repeats}, $SD_Dispatch_calledCounter-1, "JSON dispatch_repeats accuracy") if (defined($tData->{dispatch_repeats}));
-						} else { diag "SIGNALduno_Dispatch was not called, this must be an error"; }
-		            };
-		        }
-		    };
-			is($noTestRun,0,"Verify if a test was performed ");
-			$SD_Dispatch->unmock;
-		};
-		
-	};
+      TODO: {
+        local $TODO = "Checking with prerelease Version of SD_Device_ProtocolList which can fail";
+        loadJson("https://raw.githubusercontent.com/RFD-FHEM/SIGNALduino_TOOL/pre-release/FHEM/lib/SD_Device_ProtocolList.json");
+  
+        my $pID;
+        my $tID;
+      
+        my $SD_Dispatch = $mock->mock("SIGNALduno_Dispatch"); 
+        my $noTestRun=1;
+        use lib::SD_Protocols;
+        while ( ($pID, $testSet) = each  (@{$testDataArray}) )
+        {
+          next if (!lib::SD_Protocols::exists($testSet->{id}));
+          while ( ($tID, $tData) = each (@{$testSet->{data}}) ) 
+          {
+            next if (!defined($tData->{rmsg}) || !defined($tData->{dmsg}) || !defined($tData->{internals}) );
+
+            subtest "[$pID]: $testSet->{name}: [$tID] " => sub {
+              $attr{$target}{whitelist_IDs} = $testSet->{id};
+              is(AttrVal($target,"whitelist_IDs",undef), $testSet->{id}, "whitelist_IDs is ".AttrVal($target,"whitelist_IDs",undef));
```

Comment:
@sidey79, ich habe noch einen Denkfehler bzw. in den Variablen ist noch ein Fehler.
Trotz setzen des Attributes erfolgt noch keine Einschränkung auf die ID

```
        # Failed test (with amnesty) 'JSON dispatch_repeats accuracy '
        # at (eval 1003)[./FHEM/98_unittest.pm:197] line 67.
        # +-----+----+-------+
        # | GOT | OP | CHECK |
        # +-----+----+-------+
        # | 3   | eq | 7     |
        # +-----+----+-------+
```

Reply:
Grund war das Attribut maxMuMsgRepeat.
Beim damaligen erfassen der Daten wurde es auf 10 gesetzt um Nachrichten mit sehr vielen Wiederholungen besser zu erkennen.

Reply:
Sorry, PC war kaputt.

Du musst das Attribut über CommandAttr setzen. Sonst wird die attr Funktion des Modules nicht aufgerufen.

Reply:
Kein Problem.

Mein unrhytmischer Schlaf und das Wachsein brachte nun ein Endergebnis wo die Dinge gerichtet sind.

Der Test macht nun nur noch halt wegen:
```
        # SIGNALduno_Dispatch (master ID31) was not called, this must be an error
Errors of test test_DeviceData_rmsg:
"        # SIGNALduno_Dispatch (master ID31) was not called, this must be an error\n"
```
Grund:

```
ok 39 - [31]: Pollin ISOTRONIC: [0]  {
        ok 1 - whitelist_IDs is 31
        1..1
    }
```

mit den Daten
```
{"name":"Pollin ISOTRONIC", "id":"31", "module":"", "data": [
    {
      "dmsg":"u31#891FE", "user":"kaihs",
      "internals": {"DEF":"SIGNALduino_unknown_31", "NAME":"SIGNALduino_unknown_31"},
      "readings": {"state":"891FE"},
      "rmsg":"MU;P0=-12724;P1=597;P2=-667;P3=1253;P4=-1331;D=01234141412341412341414123232323232323232;CP=1;"
    }
  ]
},
```
obwohl die RMSG dispatched wird.

Bei dem Test weiß ich nur nicht wie dies zusammenhängt
```
SIGNALduino_Parse($targetHash, $targetHash, $targetHash->{NAME}, "\002".$tData->{rmsg}."\003") if (defined($tData->{rmsg}));
if ($SD_Dispatch->called() >0 )  {
....
} else { .... wo wir landen
```

Reply:
sehr seltsam. Laut logfile wird Dispatch aufgerufen ja?

Reply:
1) Es gibt auch wieder einen Unterschied zwischen Onlinetest und Local.

2) Dispatch wird dort scheinbar nicht aufgerufen, da sonst die Ausgaben
```
- SIGNALduno_Dispatch check id
- SIGNALduno_Dispatch check dmsg
```
erfolgen müssten die innerhalb der Abfrage
`if ($SD_Dispatch->called() >0 )  {` kämen.

Oder der Wert wird nicht gesetzt beim Aufruf.

Reply:
Ich kann das Verhalten bei mir lokal nachstellen. Kein Problem

Reply:
Heißt 
> Ich kann das Verhalten bei mir lokal nachstellen. Kein Problem

das du ebenso kein Problem da hast und der Test läuft?

Reply:
Folgende Startsequent ist für Protokoll 31 hinterlegt:
`start           => [1,-0.9, 1,-0.9, 1,-3.8],`

Laut Logausgabe passt diese aber nicht zu der RMSG:
`dummyDuino: Parse_MU, start pattern for MU Protocol id 31 -> LTECH not found, aborting`

Reply:
Das kontrolliere ich nochmals.
Ich habe eine masterVersion genommen und Dispatched. Das Protokoll wurde als u31 erkannt.

Reply:
masterversion von was?

Reply:
Antwort und Zusammenhang hier https://github.com/RFD-FHEM/RFFHEM/pull/796#issuecomment-588487478

Reply:
Ich verstehe nicht, was der master Stand der SD_ProtocolData.pm mit dem branch hat. :(

Protokoll 31 war mal als "Pollin" hinterlegt:
https://github.com/HomeAutoUser/RFFHEM/blame/31834c0df4cd6c5e28731dc86508bc6d72ad8b39/FHEM/lib/SD_ProtocolData.pm

Dann mit Version 1.11 https://github.com/RFD-FHEM/RFFHEM/pull/723/files wurde ID 31 frei markiert.

Jetzt ist ID 31 mit was neuem belegt:
https://github.com/HomeAutoUser/RFFHEM/blame/dev-34_new_def/FHEM/lib/SD_ProtocolData.pm

Wurde da vielleicht damals vergessen in den Testdaten auch das Protokoll 31 "frei" zu machen?

Reply:
Noch bessere Idee:
Verwende eine ID nicht neu , solange sie noch im Master verwendet wird

Reply:
Bitte nicht falsch verstehen @sidey79.
>Protokoll 31 war mal als "Pollin" hinterlegt:
https://github.com/HomeAutoUser/RFFHEM/blame/31834c0df4cd6c5e28731dc86508bc6d72ad8b39/FHEM/lib/SD_ProtocolData.pm

richtig, diese Daten sind auch so in dem JSON, stand Master

https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/7cdee1c61b1d2961becc03fad1081afa274cbbc5/FHEM/lib/SD_Device_ProtocolList.json#L357-L365

Dann wurde in der DEV Version erkannt, das das Protokoll das gleiche wie ID 34 ist. Somit wurden die Daten im pre-master geändert.

https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/fd75c4630690c5f054b8a17ba739610c2c3ecf02/FHEM/lib/SD_Device_ProtocolList.json#L600-L616

Also ist das Master JSON auch nur mit der Master Modulversion kompatibel wo auch nach meinen manuellen tests, das Protokoll 31 dispatched wurde.

>Wurde da vielleicht damals vergessen in den Testdaten auch das Protokoll 31 "frei" zu machen?

Das ist erfolgt weil im pre-release keine ID 31 mehr existent ist.

Reply:
>Noch bessere Idee:
Verwende eine ID nicht neu , solange sie noch im Master verwendet wird

Einspruch.
Bei den zeitlichen Verzügen entstehen Lücken die wir hinterher wie einen "Edammer" nicht mehr gefüllt bekommen.

DAS Problem ist doch nur, das der TEST nicht die Masterdaten nutzt wie sie im Github Master sind scheinbar. Wenn ich das ganze mit Force auf Master zurücksetzen kann, so muss es auch im Test gehen oder de rTest muss anders geschrieben werden.

Reply:
Also für mich ist der Fall klar und die Lösung auch. Für dich auch?

Reply:
> Also für mich ist der Fall klar und die Lösung auch. Für dich auch?

Ja, der Test so ist nicht eindeutig.

@Sidey ich denke nun zu wissen was das Problem ist und ich denke bei deinem Test durchzusehen.
So ist der Test nicht aussagekräftig weil ich den Stand "Trabant" mit "Mercedes" vergleichen möchte *smile*

- Master Daten JSON können nur mit einem Dispatch von der Masterversion verglichen werden.
- pre-release Daten JSON können nur mit einem Dispatch von der DEVversion verglichen werden.

Die Daten von dem Master JSON bestand werden geprüft mit der pre-releaseVersion.
Weil die ID 31 damals nun nicht mit der RAWMSG arbeitet ist klar. Diese ID wurde bereits neu verarbeitete mit Daten die einen anderen Start haben.

Wir können die Daten so nicht vergleichen. Der Test wäre erst plausibel, wenn die JSDON Daten auch mit den Daten vom Master geprüft werden. Alles andere bringt solche Fehlerfälle, welceh eigendlich keine sind. Es sind Korrekturen welche aufgedeckt wurden.

Was soll deine Lösung bzw. dein Vorschlag sein?

Reply:
@sidey79, mein Vorschlag ist wie folgt.

Da wir vermutlich nicht die Testumgebung
```
- Master Daten JSON können nur mit einem Dispatch von der Masterversion verglichen werden.
- pre-release Daten JSON können nur mit einem Dispatch von der DEVversion verglichen werden.
```
generieren können, so wird der Test erweitert oder wir winken es derzeit so durch. Es dient der Ordnung und Protokolle und wir sind dort ja schon zu 95% durch.

a) durchwinken oder ein TODO ohne Stop einbauen
b) Test erweitern wenn so ein Fall auftreten sollte, zu prüfen ob überhaupt ein Dispatch erfolgt, wenn ja ,ist es plausibel (WEIL wir mit dem DEV Mpodul prüfen und nicht Master Modul)
c) Test nur auf die aktuellen DEV Daten beziehen und nicht die JSON Master Daten gegenprüfen.
d) Ich entferne die Daten aus dem Master weil es Daten eines uProtokolles waren

Reply:
Wieso nicht

e) Eine Protokoll ID ist so lange für eine Neuverwendung gesperrt, biss es auch im Master nicht mehr verwendet wird.

Das fände ich grundsätzlich eine Regel die derartige Probleme verhindert.

Reply:
Du weißt selbst wie "aktuell" wir momentan die Daten ins Master schieben.
Wenn wir nun über ein Jahr wieder eine "Sperrfrist" haben weil UpdateIntervall sehr sporadisch ist, so kommt es vor, das wir wieder Lücken uns künstlich schaffen wo wir hinterher nach der Zeit vielleicht nicht mehr wissen wie der Zusammenhang war. Wir dokumentiren viel aber nicht alles. Ich bin mir Sicher @elektron-bbs würde genau so denken.

Da ich in dem Datenbestand eine Nachricht dokumentierte, welche als u und nur halb fertig war, so kam dies zu dem "Fehler". Ich habe diese Nachricht aus dem JSON entfernt weil sie eh beim nächsten Update von pre-release to master frisch und richtig eingearbeitet wird.

Der Test läuft nun Fehlerfrei durch. Die Vorschrift benötigen wir nicht, es ist einfach "
"Übereifer" von mir gewesen ;-)

... Das nun ein anderer Test fehl schlägt .... 

`Starting test test_sub_SIGNALduino_Parse_MU-definition.txt`
Hinweis schon im TEST
`note("Test can be extended to JSON MU data");`

Das sehe ich fast wichtiger zu kären, weil dieser ja eigendlich doppelt ist.
Das Gute, die Fehler zeigen nun, das wir zu viel doppelt auch prüfen :-)